src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,443 @@
+        dist_to_start = rel_next_position - rel_start_sample
+
+        # get the start and end position inside the ringbuffer
+        end_pos = (
+            self._timestamp_to_rel_index(rb_current_position) + 1
+        ) - dist_to_start
+
+        # Note that these check is working since we are using the positions
+        # relative to the oldest sample stored in the MovingWindow.
+        if rel_start_sample <= rel_next_position < rel_end_sample:
+            # end position is start_position of the window that is currently written
+            # that's how end_pos is currently set
+            log.debug("Next sample will be inside the window time interval!")
+        else:
+            log.debug("Next sample will be outside the window time interval!")
+            # end position is start_position of the window that
+            # is overwritten next, hence we adding distance.
+            end_pos += self._distance
+
+        # add the offset to the oldest sample in the ringbuffer and wrap around
+        # to get the start and end positions in the ringbuffer
+        rb_offset = self._datetime_to_index(self._moving_window.time_bound_oldest())
+        start_pos = self._buffer.wrap(end_pos + self._distance + rb_offset)
+        end_pos = self._buffer.wrap(end_pos + rb_offset)
+
+        log.debug("start_pos in ringbuffer: %s", start_pos)
+        log.debug("end_pos in ringbuffer: %s", end_pos)
+
+        return (start_pos, end_pos, window_size)
+
+    def _get_reshaped_np_array(
+        self, start: datetime, end: datetime
+    ) -> Tuple[NDArray[np.float_], int]:
+        """
+        Create a reshaped numpy array from the MovingWindow.
+
+        The reshaped array is a two dimemsional array, where one dimension
+        is the window_size and the other the number of windows returned by
+        the `_get_buffer_bounds` method.
+
+        Args:
+            start: The start timestamp of the window.
+            end: The end timestamp of the window.
+
+        Returns:
+            A tuple containing the reshaped numpy array and the window size.
+        """
+        (start_pos, end_pos, window_size) = self._get_buffer_bounds(start, end)
+
+        if start_pos >= end_pos:
+            window_start = self._moving_window[start_pos:]
+            window_end = self._moving_window[:end_pos]
+            # make the linter happy
+            assert isinstance(window_start, np.ndarray)
+            assert isinstance(window_end, np.ndarray)
+            window_array = np.concatenate((window_start, window_end))
+        else:
+            # ignoring the type check because we know that the returned type is an ndarray
+            window_array = self._moving_window[start_pos:end_pos]  # type: ignore[assignment]
+
+        return (self._reshape_np_array(window_array, window_size), window_size)
+
+    def avg(
+        self, start: datetime, end: datetime, weights: List[float] | None = None
+    ) -> NDArray[Any]:
```

Comment:
Shouldn't this return `NDArray[np.float_]`? I think it's not that terrible to do some type ignore in the internal methods but I would try to keep the public interface as strictly typed as possible.